rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp

```diff
@@ -76,26 +79,78 @@ class PublicationManager
 
   void run_publishers(bool verbose = false) const
   {
-    for (const auto & pub_fcn : publishers_) {
+    for (const auto & pub_fcn : publishers_fcn_) {
       pub_fcn(verbose);
     }
+    rclcpp::executors::SingleThreadedExecutor exec;
+    exec.spin_node_some(pub_node_);
+    std::this_thread::sleep_for(100ms);
   }
 
   void run_publishers_async(bool verbose = false) const
   {
     std::vector<std::future<void>> futures;
-    for (const auto & pub_fcn : publishers_) {
+    for (const auto & pub_fcn : publishers_fcn_) {
       futures.push_back(std::async(std::launch::async, pub_fcn, verbose));
     }
     for (auto & publisher_future : futures) {
       publisher_future.get();
     }
   }
 
+  /// \brief Wait until publisher with specified topic will be connected to the subscribers or
+  ///    timeout occur.
+  /// \tparam Timeout Data type for timeout duration from std::chrono:: namespace
+  /// \param topic_name topic name to find corresponding publisher
+  /// \param timeout Maximum time duration during which discovery should happen.
+  /// \param n_subscribers_to_match Number of subscribers publisher should have for match.
+  /// \return true if not find publisher by topic name or publisher has specified number of
+  ///    subscribers, otherwise false.
+  template<typename Timeout>
+  bool spin_and_wait_for_matched(
+    const char * topic_name,
+    Timeout timeout, size_t n_subscribers_to_match = 1)
+  {
+    rclcpp::PublisherBase::SharedPtr publisher = nullptr;
+    for (const auto pub : publishers_) {
+      if (!std::strcmp(pub->get_topic_name(), topic_name)) {
+        publisher = pub;
+        break;
+      }
+    }
+
+    if (publisher == nullptr) {
+      return false;
+    }
+
+    if (publisher->get_subscription_count() +
+      publisher->get_intra_process_subscription_count() >= n_subscribers_to_match)
+    {
+      return true;
+    }
+
+    using clock = std::chrono::steady_clock;
+    auto start = clock::now();
+
+    rclcpp::executors::SingleThreadedExecutor exec;
+    do {
+      exec.spin_node_some(pub_node_);
+
+      if (publisher->get_subscription_count() +
+        publisher->get_intra_process_subscription_count() >= n_subscribers_to_match)
+      {
+        return true;
+      }
+    } while ((clock::now() - start) < timeout);
+
+    return false;
+  }
+
 private:
   std::shared_ptr<rclcpp::Node> pub_node_;
   using PublicationF = std::function<void (bool)>;
-  std::vector<PublicationF> publishers_;
+  std::vector<PublicationF> publishers_fcn_;
```

Comment:
Is there any reason why we should maintain two separate vectors? I would prefer 

```
std::vector<std::pair<rclcpp::PublisherBase::SharedPtr, PublicationF>> publishers_;
```

This way it's impossible for the vectors to get out of sync.